plugin/core/main.py

```diff
@@ -0,0 +1,300 @@
+        env[var] = os.path.expandvars(sublime.expand_variables(value, variables))
+
+    client = start_server(expanded_args, project_path, env)
+    if not client:
+        window.status_message("Could not start " + config.name + ", disabling")
+        debug("Could not start", config.binary_args, ", disabling")
+        return None
+
+    initializeParams = {
+        "processId": client.process.pid,
+        "rootUri": filename_to_uri(project_path),
+        "rootPath": project_path,
+        "capabilities": {
+            "textDocument": {
+                "completion": {
+                    "completionItem": {
+                        "snippetSupport": True
+                    }
+                },
+                "synchronization": {
+                    "didSave": True
+                }
+            },
+            "workspace": {
+                "applyEdit": True
+            }
+        }
+    }
+    if config.init_options:
+        initializeParams['initializationOptions'] = config.init_options
+
+    client.send_request(
+        Request.initialize(initializeParams),
+        lambda result: handle_initialize_result(result, client, window, config))
+    return client
+
+
+def start_window_client(view: sublime.View, window: sublime.Window, config: ClientConfig) -> Client:
+
+    clients = window_clients(window)
+    if config.name not in clients:
+        client = start_client(window, config)
+    else:
+        client = clients[config.name]
+
+    return client
+
+
+def start_server(server_binary_args, working_dir, env):
+    debug("starting " + str(server_binary_args))
+    si = None
+    if os.name == "nt":
+        si = subprocess.STARTUPINFO()  # type: ignore
+        si.dwFlags |= subprocess.SW_HIDE | subprocess.STARTF_USESHOWWINDOW  # type: ignore
+    try:
+        process = subprocess.Popen(
+            server_binary_args,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            cwd=working_dir,
+            env=env,
+            startupinfo=si)
+        return Client(process, working_dir)
+
+    except Exception as err:
+        sublime.status_message("Failed to start LSP server {}".format(str(server_binary_args)))
+        exception_log("Failed to start server", err)
+
+
+def apply_workspace_edit(window, params):
+    edit = params.get('edit')
+    window.run_command('lsp_apply_workspace_edit', {'changes': edit})
+
+
+class LspRestartClientCommand(sublime_plugin.TextCommand):
+    def is_enabled(self):
+        return is_supported_view(self.view)
+
+    def run(self, edit):
+        window = self.view.window()
+        unload_window_clients(window.id())
+
+
+class LspStartClientCommand(sublime_plugin.TextCommand):
+    def is_enabled(self):
+        return is_supported_view(self.view)
+
+    def run(self, edit):
+        start_active_view()
+
+
+class LspApplyWorkspaceEditCommand(sublime_plugin.WindowCommand):
```

Comment:
Would it make sense to move this and the following class to plugin/edit.py or plugin/apply_edit.py?

It looks like a normal command such as Symbols / Diagnostics / Code Actions ...

Reply:
Good point, might as well move it out of main!
 I'd like to have part of the edit functionality in the core package because it's much more a service to other features rather than something the user invokes directly.